### ISSUE TITLE
Proposal: Way to express pagination/cursors of "link" objects

### ISSUE THREAD

**Issue by butsjoh** [json-api/json-api#212]
We have a use case here where we are trying to express pagination for links (associations) of a document. In the extending section of the json api website there is clearly stated that you could express pagination through the meta section of the document. It does not say anything on how that would be possible for links (associations). 

> The documentation of a "link" object says the following:
> 
> The link to each related resource MUST be one of the following:
> - a string or number - to represent a single ID.
> - an array of strings or numbers - to represent multiple IDs.
> - a "link" object that contains one or more of the attributes: "id", "ids", "href" and "type". Note that "id" and "ids" should never be present together.
> - an array of "link" objects

What we are proposing is to allow an additional key for a link object to express more information about the link which can then be used as well to express pagination or other metadata about the link. Naming for that key could be either "meta" or maybe even "cursors", it all depends what this would mean for the spec.

Example:

Main document is a conversation which has an association to messages. We are limiting the included set of messages to 3. But expressing the next set to load in the meta of the link as the cursors example.

``` json
{
    "conversations": [
        {
            "id": "99e47e8f-1e6c-4d00-9259-82855296ae99",
            "href": "http://example.com/messaging/conversations/99e47e8f-1e6c-4d00-9259-82855296ae99",
            "subject": "How was your day?",
            "created_at": "2013-05-21T00:00:00Z",
            "updated_at": "2013-05-22T00:00:00Z",
            "links": {
                "participants": {
                    "ids": [
                        "81bd6f56-41dc-419f-89b5-c9a3d53e010e",
                        "8f468f75-1332-42d2-8f3f-aa79aab0ebe2"
                    ],
                    "href": "http://example.com/api/resources/?ids={conversations.participants}",
                    "type": "resource"
                },
                "messages": {
                    "ids": [
                        "f49dc39a-ee4c-4d06-8d8e-07d7ec8ad802",
                        "81c64349-e748-411c-9549-5529bc7199bb",
                        "71c64349-e748-411c-9549-5529bc717463"
                    ],
                    "href": "http:/example.com/api/messaging/messages?conversation_id=99e47e8f-1e6c-4d00-9259-82855296ae99",
                    "type": "message",
                    "meta": {
                        "after": "ge34c39a-ee4c-4d06-8d8e-07d7ec8a2922",
                        "before": "",
                        "limit": 3,
                        "href": "http://example.com/api/messaging/messages/{conversation_id,after,before,limit}"
                    }
                }
            }
        }
    ],
    "linked": {
        "resources": [
            {
                "id": "81bd6f56-41dc-419f-89b5-c9a3d53e010e",
                "href": "http://example.com/api/resources/81bd6f56-41dc-419f-89b5-c9a3d53e010e"
            },
            {
                "id": "8f468f75-1332-42d2-8f3f-aa79aab0ebe2",
                "href": "http://example.com/api/resources/8f468f75-1332-42d2-8f3f-aa79aab0ebe2"
            }
        ],
        "messages": [
            {
                "id": "f49dc39a-ee4c-4d06-8d8e-07d7ec8ad802",
                "href": "http://example.com/api/messaging/messages/f49dc39a-ee4c-4d06-8d8e-07d7ec8ad802",
                "body": "Was wondering how you are doing today."
            },
            {
                "id": "81c64349-e748-411c-9549-5529bc7199bb",
                "href": "http://example.com/api/messaging/messages/81c64349-e748-411c-9549-5529bc7199bb",
                "body": "Doing really well"
            }
        ]
    },
    "meta": {}
}
```

So the line in the spec about the keys of the link object would change:

from 

> a "link" object that contains one or more of the attributes: "id", "ids", "href" and "type". Note that "id" and "ids" should never be present together.

to

> a "link" object that contains one or more of the attributes: "id", "ids", "href", "type" and **"meta"**. Note that "id" and "ids" should never be present together.

For me the key name does not have to be meta but i would like to have a way to express this. 

Other possibilities are adding it to the meta of the main document

``` json
{
    "conversations": [
        {}
    ],
    "meta": {
        "messages": {
            "after": "ge34c39a-ee4c-4d06-8d8e-07d7ec8a2922",
            "before": "",
            "limit": 3,
            "href": "http://example.com/api/messaging/messages/{conversation_id,after,before,limit}"
        }
    }
}
```

 but that will fail if you would load a collection of the same documents (in this case multiple conversations)

``` json
{
    "conversations": [
        {}, {}, {}, {}
    ],
    "meta": {
       "messages": {
         // What todo now? create a key for each conversation and then repeat messages stuff?  
       }
   }
}
```

So if the proposal to allow some sort of key to the "link" object to express additional data seems valid i can create a pull request to update the documentation. Let me hear your opinions :)

**Comment by Bertg** [comment]
To add to this - OP is a colleague of mine - I think that the spec should relax its "MUST" assertion on link objects. Namely it should allow a API creator to add keys to it that can indicate pagination information. 

Say if you use the "cursor" key to paginate the main collection, you should be allowed to use "cursor" key in the linked object as well.

**Comment by butsjoh** [comment]
@dgeb Any opinion/suggestion/remark ? :)

**Comment by dgeb** [comment]
Since the general pattern is that specific link objects can override top level settings if necessary, I would support allowing meta objects within individual link objects.

I would also like the spec to at least provide guidance for both offset/limit and cursor based pagination.

**Comment by butsjoh** [comment]
@dgeb Awesome :) Will try to create a pull request that will explain this idea. We can then discuss again on the details.

**Comment by mattolson** [comment]
+1 on this. We also want to support paging for linked resources and this seems the most sensible. The alternative is to put all this information in the top level meta block, but it makes more sense to put it in context.

**Comment by ahacking** [comment]
I think paging could be handled elegantly with my [radical proposal at the end of #276](https://github.com/json-api/json-api/issues/276#issuecomment-61397312)

Basically json-api need to use link templates more earnestly and then it can be trimmed down and be silent on so many things, including ids, urls, resource structure, polymorphism, resource collections, results vs linked data, embedded resources and indeed paging.

**Comment by natbusa** [comment]
i like the idea of links pagination. I am not sure about the fields inside meta, but that's something else. For now i am fine to keep it format free.

**Comment by dgeb** [comment]
Pagination in general is now covered in the spec: http://jsonapi.org/format/#fetching-pagination

Pagination of relationships should be included in #348.
